libvast/src/table_slice_builder.cpp

```diff
@@ -33,6 +52,136 @@ table_slice_builder::~table_slice_builder() noexcept {
 
 // -- properties ---------------------------------------------------------------
 
+size_t table_slice_builder::columns() const noexcept {
+  auto result = schema_->num_fields();
+  VAST_ASSERT(result >= 0);
+  return detail::narrow_cast<size_t>(result);
+}
+
+namespace {
+
+/// Create a table slice from a record batch.
+/// @param record_batch The record batch to encode.
+/// @param builder The flatbuffers builder to use.
+/// @param serialize Embed the IPC format in the FlatBuffers table.
+table_slice
+create_table_slice(const std::shared_ptr<arrow::RecordBatch>& record_batch,
+                   flatbuffers::FlatBufferBuilder& builder, type schema,
+                   table_slice::serialize serialize) {
+  VAST_ASSERT(record_batch);
+#if VAST_ENABLE_ASSERTIONS
+  // NOTE: There's also a ValidateFull function, but that always errors when
+  // using nested struct arrays. Last tested with Arrow 7.0.0. -- DL.
+  auto validate_status = record_batch->Validate();
+  VAST_ASSERT(validate_status.ok(), validate_status.ToString().c_str());
+#endif // VAST_ENABLE_ASSERTIONS
+  auto fbs_ipc_buffer = flatbuffers::Offset<flatbuffers::Vector<uint8_t>>{};
+  if (serialize == table_slice::serialize::yes) {
+    auto ipc_ostream = arrow::io::BufferOutputStream::Create().ValueOrDie();
+    auto opts = arrow::ipc::IpcWriteOptions::Defaults();
+    opts.codec
+      = arrow::util::Codec::Create(
+          arrow::Compression::ZSTD,
+          arrow::util::Codec::DefaultCompressionLevel(arrow::Compression::ZSTD)
+            .ValueOrDie())
+          .ValueOrDie();
```

Comment:
Maybe add a comment to make this configurable.

Reply:
I think we can actually remove it. This was only necessary for the segment store, Feather and Parquet already have compression configurable on their own.